### PR TITLE
Make expect tests more usable

### PR DESCRIPTION
**Pull request by v-gb**
One big problem with expect tests right now is that the rule building the expect tests runner declares almost none of its dependencies, forcing people to clean to run such tests when they change the compiler. The first commit fixes this.

Another problem, though maybe it's because I don't understand the makefiles, is that there seems to be no way to just run a test without first knowing everything the test will need and building them. Of course, one can overapproximate by building world or something, but that's slower than necessary. So the second commit introduces a phony target to build what's necessary to run an expect test and no more.
Once `configure` has been called, one can run a test like this: `make -C testsuite/tests/typing-sigsubst DIFF="diff -u" deps default`. This command takes 1min the first time, .2s when changing the test, 2s when changing typemod.ml, 3.5s when changing parser.mly.
Using `make world && make -C testsuite tools && make -C testsuite/tests/typing-sigsubst DIFF="diff -u" default` takes 2min the first time, the last call to `make` takes .15s when changing the test, and the whole command takes 10s when changing typemod.ml, and 20s when changing parser.mly.

### CODE REVIEW

**Review by gasche — APPROVED** [review]
This looks to me like a safe change -- preserving current use-cases, and un-breaking new use-cases.

**Comment by shindere** [comment]
Gabriel Scherer (2017/09/17 06:10 +0000):
> gasche approved this pull request.
> 
> This looks to me like a safe change -- preserving current use-cases,
> and un-breaking new use-cases.

Well, I actually already migrated the expect tests to using ocamltest so
I am wondering whether this PR wouldn't, at least partially, become
obsoleted by the introduction of ocamltest. Would it be possible to wait
till I can merge the migrated way of running these tests and then see
what in this PR remains useful?

**Comment by v-gb** [comment]
The first commit is useful regardless of how the tests are run.

For the second commit, ocamltest doesn't solve the problem that there's no good way to just run a test including building what's necessary to run said test. It probably makes the problem worse by being one more tool that needs to be built. The specifc way I solved this problem wouldn't work anymore with ocamltest, though.
But in any case, is #681 actually ready to run expect tests? I couldn't find a single translated expect test, nor an equivalent to `make promote`.

**Comment by shindere** [comment]
sliquister (2017/09/17 07:37 -0700):
> The first commit is useful regardless of how the tests are run.
> 
> For the second commit, ocamltest doesn't solve the problem that
> there's no good way to just run a test including building what's
> necessary to run said test. It probably makes the problem worse by
> being one more tool that needs to be built.

It's true that ocamltest itself does not take care of building e.g. the
expect program itself. That being said, I don't think it makes things
worse because #681 makes sure ocamltest.byte gets built during make
world while ocamltest.opt gets build during make world.opt.

> But in any case, is #681 actually ready to run expect tests?

Not in itself.

> I couldn't find a single translated expect test, nor an equivalent to
> `make promote`.

You're right. Actually, I did migrate the tests one year ago so #681
used to contain that work. But then I couldn't take care of this work
for some time and when I could work on it again, I did a rebase and
started to rewrite the history to have something cleaner than what I had
at first. The expect tests have not been re-migrated but since I was
able to migrate them last year I hope to be able to submit a PR
migrating them soon after this one has been merged.

**Comment by gasche** [comment]
Let's wait until ocamltest is merged to see what the situation is. The present PR is small and is easily adapted/rebased, and I don't want to make more test-related work or worries for @shindere before the merge.

**Comment by shindere** [comment]
Gabriel Scherer (2017/09/17 13:10 -0700):
> Let's wait until ocamltest is merged to see what the situation is. The
> present PR is small and is easily adapted/rebased, and I don't want to
> make more test-related work or worries for @shindere before the merge.

Thanks. Tomorrow, I hope. Well I guess it wouldn't hurt to merge this PR
if this is what is desired. It's just that it may make the re-migration
of expect based tests a bit less straightforward than a mere rebase...

**Comment by v-gb** [comment]
Now that ocamltest was merged, can we consider merging this?

**Comment by gasche** [comment]
Merging it now works for me. @shindere?

**Review by gasche — APPROVED** [review]

**Comment by shindere** [comment]
sliquister (2017/11/19 22:42 +0000):
> Now that ocamltest was merged, can we consider merging this?

It's on my list. I just would like to finish the hooks thing and then
re-migrate the tests already migrated last year, which include the
expect tests. Will comment here ASAP, sorry for the delay.

**Comment by shindere** [comment]
Gabriel Scherer (2017/11/20 05:56 +0000):
> gasche approved this pull request.

@gasche if you want to merge that's fine with me, actually. I'll adapt
the tests to the current situation while re-migrating, if this is
necessary.

**Comment by gasche** [comment]
I'll merge because the change is small and I prefer to avoid having external contributions wait for too long. Thanks everyone!

**Pull request merged**
